### ISSUE TITLE
Nested Lopping ( double looping)

### ISSUE THREAD

**Issue by Vadims06** [hashicorp/terraform#20288]
### Current Terraform Version

terraform -v
Terraform v0.11.11
+ provider.aws v1.57.0
...

### Use-cases
I would like to add multiple policies into one user group. User groups are several. For example i defined the following map with user groups as keys and group policies as a values:

```hcl
groups_w_policies = {
"admins": ["RW_EC2_access_policy", "RW_Network_policy"],
"users": ["RO_EC2_access_policy", "RO_Network_policy"],
"qa": ["RO_EC2_access_policy", "Forse_changing_pass_policy"] }
```

In order to attach group policies into user group, I have to go over all keys [i mean user groups] and go over nested routine through group policies. 
But it seems that it impossible, am I right?

### Attempted Solutions

```hcl
resource "aws_iam_group_policy_attachment" "test-attach" {
  count = "${length(keys(var.groups_w_policies))}"
  group      = "${element(keys(var.groups_w_policies), count.index)}"
  count_policies = "${length(values(var.groups_w_policies))}"
  policy_arn = "${element(element(values(var.groups_w_policies), count.index), count_policies.index)}"
}
```

**Comment by apparentlymart** [comment]
Hi @Vadims06!

I think you're right that this is not straightforward to achieve in current releases of Terraform, and would probably require a different structure for `groups_w_policies` to make it possible to iterate it in only one step:

```hcl
  groups_w_policies = [
    ["admins", "RW_EC2_access_policy"],
    ["admins", "RW_Network_policy"],
    ["users", "RO_EC2_access_policy"],
    ["users", "RO_Network_policy"],
    ["qa", "RO_EC2_access_policy"],
    ["qa", "Force_changing_pass_policy"],
  ]
```

In the forthcoming v0.12.0 release of Terraform it should be possible to automatically transform your original structure into this flattened form so that it can be iterated in a single pass while still being easy to maintain. For example:

```hcl
  # Requires Terraform v0.12
  groups_w_policies = {
    admins = ["RW_EC2_access_policy", "RW_Network_policy"],
    users  = ["RO_EC2_access_policy", "RO_Network_policy"],
    qa     = ["RO_EC2_access_policy", "Forse_changing_pass_policy"],
  }
  groups_w_policies_flat = flatten([
    for group, policies in local.groups_w_policies : [
      for policy in policies : {
        group  = group,
        policy = policy,
      }
    ]
  ])
```

The second expression here means that you should be able to use `local.groups_w_policies_flat` as your iteration basis, with each element giving you one (group, policy) pair.

```hcl
resource "aws_iam_group_policy_attachment" "test-attach" {
  count = length(local.groups_w_policies_flat)

  group      = local.groups_w_policies_flat[count.index].group
  policy_arn = local.policy_arns[local.groups_w_policies_flat[count.index].policy]
}
```

Once #17179 is implemented (planned for a minor release after v0.12.0) this could be expressed in a different way that avoids dealing with list indices:

```hcl
resource "aws_iam_group_policy_attachment" "test-attach" {
  for_each = flatten([
    for group, policies in local.groups_w_policies : [
      for policy in policies : {
        group  = group,
        policy = policy,
      }
    ]
  ])

  group      = each.value.group
  policy_arn = local.policy_arns[each.value.policy]
}
```

Since the [`for` expression](https://github.com/hashicorp/terraform/blob/5f8916b4fd851e85cdf16cfcd5c635d1520b419e/website/docs/configuration/expressions.html.md#for-expressions) feature is already implemented in master ready to be included in the release, and the resource `for_each` feature is already covered by #17179, I'm going to close this out just to keep the discussion consolidated.

In the mean time, the workaround I described above of manually flattening the list is probably your best bet, until v0.12.0 is released and you are ready to migrate to it.

**Comment by Vadims06** [comment]
@apparentlymart Thank you for your reply. Actually, I made such construction
```
groups_and_policies = [
["test_group1", "arn:aws:iam::aws:policy/job-function/SupportUser"],
["test_group1", "arn:aws:iam::aws:policy/AmazonEC2ReadOnlyAccess"],
["test_group2", "arn:aws:iam::aws:policy/AmazonEC2ReadOnlyAccess"],
]
```
and tried to use it in aws_iam_group_policy_attachment function, but it doesn't work. Am i right in my attemp to retriev nested lists and then to get the first and zerost element from the nested list?
```
resource "aws_iam_group_policy_attachment" "test-attach" {
  count = "${length(var.groups_and_policies)}"
  policy_arn  = "${ element( element(var.groups_and_policies, count.index), 1)}"
  group = "${element( element(var.groups_and_policies, count.index), 0)}"
}
```
The error
```
Error: aws_iam_group_policy_attachment.test-attach: 3 error(s) occurred:

* aws_iam_group_policy_attachment.test-attach[0]: At column 4, line 1: element: argument 1 should be type list, got type string in:

${ element( element(var.groups_and_policies, count.index), 1)}
* aws_iam_group_policy_attachment.test-attach[1]: At column 4, line 1: element: argument 1 should be type list, got type string in:

${ element( element(var.groups_and_policies, count.index), 1)}
* aws_iam_group_policy_attachment.test-attach[2]: At column 4, line 1: element: argument 1 should be type list, got type string in:

${ element( element(var.groups_and_policies, count.index), 1)}
```
